crates/client/src/keyring/mod.rs

```diff
@@ -0,0 +1,312 @@
+//! Utilities for interacting with keyring and performing signing operations.
+
+use crate::config::Config;
+use crate::RegistryUrl;
+use indexmap::IndexSet;
+use secrecy::Secret;
+use warg_crypto::signing::PrivateKey;
+
+mod error;
+use error::KeyringAction;
+pub use error::KeyringError;
+
+/// Interface to a pluggable keyring backend
+#[derive(Debug)]
+pub struct Keyring {
+    imp: Box<keyring::CredentialBuilder>,
+    name: &'static str,
+}
+
+/// Result type for keyring errors.
+pub type Result<T, E = KeyringError> = std::result::Result<T, E>;
+
+impl Keyring {
+    #[cfg(target_os = "linux")]
+    /// List of supported credential store backends
+    pub const SUPPORTED_BACKENDS: &'static [&'static str] =
+        &["secret-service", "linux-keyutils", "mock"];
+    #[cfg(any(target_os = "freebsd", target_os = "openbsd"))]
+    /// List of supported credential store backends
+    pub const SUPPORTED_BACKENDS: &'static [&'static str] = &["secret-service", "mock"];
+    #[cfg(target_os = "windows")]
+    /// List of supported credential store backends
+    pub const SUPPORTED_BACKENDS: &'static [&'static str] = &["windows", "mock"];
+    #[cfg(target_os = "macos")]
+    /// List of supported credential store backends
+    pub const SUPPORTED_BACKENDS: &'static [&'static str] = &["macos", "mock"];
+    #[cfg(target_os = "ios")]
+    /// List of supported credential store backends
+    pub const SUPPORTED_BACKENDS: &'static [&'static str] = &["ios", "mock"];
+    #[cfg(not(any(
+        target_os = "linux",
+        target_os = "freebsd",
+        target_os = "openbsd",
+        target_os = "macos",
+        target_os = "ios",
+        target_os = "windows",
+    )))]
+    /// List of supported credential store backends
+    pub const SUPPORTED_BACKENDS: &'static [&'static str] = &["mock"];
+
+    /// The default backend when no configuration option is set
+    pub const DEFAULT_BACKEND: &'static str = Self::SUPPORTED_BACKENDS[0];
+
+    /// Returns a human-readable description of a keyring backend.
+    pub fn describe_backend(backend: &str) -> &'static str {
```

Comment:
Would it be difficult to make `backend` an enum?

Reply:
I considered this but decided it would create more problems than it would solve. The set of supported backends is platform-dependent, so there would be a distinction between the enumeration of all understood backend identifiers, and the subset of those which are supported on the user's platform. So there would be a design trade-off between:

1. Defining two separate enums, one which covers all backends and one which covers those which are supported on the platform.
2. Defining only the supported enum, and needing a bunch of conditional attributes every time you want to match on it.
3. Defining only the all-backends enum, and putting an `is_supported(&self) -> bool` method on it.

Then we have to decide how exactly we want our `Serialize` and `Deserialize` implementations to work and which type they should use, and probably put custom error reporting into this traits to not be too cryptic if the user copies a configuration file from one system to another where the backend isn't supported. Altogether it didn't seem worth it, so I just kept this stringly typed.